qubes/features.py

```diff
@@ -0,0 +1,189 @@
+                value=value)
+
+    def clear(self):
+        for key in tuple(self):
+            del self[key]
+
+    def pop(self, _key, _default=None):
+        '''Not implemented
+        :raises: NotImplementedError
+        '''
+        raise NotImplementedError()
+
+    def popitem(self):
+        '''Not implemented
+        :raises: NotImplementedError
+        '''
+        raise NotImplementedError()
+
+    def setdefault(self, _key, _default=None):
+        '''Not implemented
+        :raises: NotImplementedError
+        '''
+        raise NotImplementedError()
+
+    def update(self, other=None, **kwargs):
+        if other is not None:
+            if hasattr(other, 'keys'):
+                for key in other:
+                    self[key] = other[key]
+            else:
+                for key, value in other:
+                    self[key] = value
+
+        for key in kwargs:
+            self[key] = kwargs[key]
+
+    #
+    # end of overriding
+    #
+
+    def _recursive_check(self, attr=None, *, feature, default,
+            check_adminvm=False, check_app=False):
+        '''Recursive search for a feature.
+
+        Traverse domains along one attribute, like
+        :py:attr:`qubes.vm.qubesvm.QubesVM.netvm` or
+        :py:attr:`qubes.vm.appvm.AppVM.template`, starting with current domain
+        (:py:attr:`subject`). Search stops when first `feature` is found. If
+        a qube has no attribute, or if the attribute is :py:obj:`None`, the
+        *default* is returned, or if not specified, :py:class:`KeyError` is
+        raised.
+
+        If `check_adminvm` is true, before returning default, also AdminVM is
+        consulted (the recursion does not restart).
+
+        If `check_app` is true, also the app feature is checked. This is not
+        implemented, as app does not have features yet.
+        '''
+        if check_app:
+            raise NotImplementedError('app does not have features yet')
+
+        assert isinstance(self.subject, _vm.BaseVM), (
+            'recursive checks do not work for {}'.format(
+                type(self.subject).__name__))
+
+        try:
+            return self[feature]
+        except KeyError:
+            pass
+
+        if attr is not None:
+            next_subject = getattr(self.subject, attr, None)
+            assert not next_subject is self.subject, 'infinite recurse'
+            if next_subject is not None:
+                # pylint: disable=protected-access
+                return next_subject.features._recursive_check(attr,
+                    feature=feature, default=default,
+                    check_adminvm=check_adminvm, check_app=check_app)
+
+        if check_adminvm:
+            adminvm = self.subject.app.domains['dom0']
+            if adminvm not in (None, self.subject):
+                try:
+                    return adminvm.features[feature]
+                except KeyError:
+                    pass
+
+        if check_app:
```

Comment:
Since it doesn't work yet (and in fact is broken, not only because app does not have `features` attribute), please remove it or at least comment it out. You may leave a `TODO` comment here.